fletcher/_algorithms.py

```diff
@@ -336,3 +337,54 @@ def _merge_valid_bitmaps(a: pa.Array, b: pa.Array) -> np.ndarray:
         )
 
         return result
+
+
+@njit(fastmath=True)
+def _get_new_indptr(self_indptr, indices, new_indptr):
+    for i in range(len(indices)):
+        row = indices[i]
+        new_indptr[i + 1] = new_indptr[i] + self_indptr[row + 1] - self_indptr[row]
+    return new_indptr
+
+
+@njit(fastmath=True, parallel=True)
+def _fill_up_indices(new_indptr, new_indices, self_indices, self_indptr, indices):
+    for i in prange(len(indices)):
+        row = indices[i]
+        size = self_indptr[row + 1] - self_indptr[row]
+        if size > 0:
+            new_indices[new_indptr[i] : new_indptr[i + 1]] = self_indices[
+                self_indptr[row] : self_indptr[row + 1]
+            ]
+
+
+def take_on_pyarrow_list(array, indices):
+    """Return a pyarrow.ListArray or pyarrow.LargeListArray containing only the rows of the given indices."""
+    if len(array.flatten()) == 0:
+        return array.take(pa.array(indices))
+
+    dtype = np.int64 if pa.types.is_large_list(array.type) else np.int32
```

Comment:
this logic will overflow in the case where there's a lots of indices even if the initial dtype int32.